common/core/web/input-processor/test.sh

```diff
@@ -54,48 +54,39 @@ if [ $FETCH_DEPS = true ]; then
 fi
 
 # Ensures that the lexical model compiler has been built locally.
-#echo_heading "Preparing Lexical Model Compiler for test use"
-#pushd $WORKING_DIRECTORY/node_modules/@keymanapp/lexical-model-compiler
-#npm run build
-#popd
+echo_heading "Preparing Lexical Model Compiler for test use"
+pushd "$KEYMAN_ROOT/developer/js/"
+npm run build
```

Comment:
```suggestion
./build.sh
```

`./build.sh` to keep it consistent with our build processes? Also, this should be optional -- we shouldn't force a rebuild for a test run... But perhaps I'll resolve this particular aspect with the build.sh refactoring coming soon (which moves test.sh into `build.sh test`, where a build would happen with `build.sh build test`.

Reply:
> we shouldn't force a rebuild for a test run

But we should force a build when it's never been built at all.  Which is what we'd expect to see on a CI build, no?  Builds of `input-processor` (for Web) are separate of those for the lexical model compiler (for Developer).

Might be nice if we had some clean and simple "build if unbuilt" command.  Do we have any existing patterns for that?

Reply:
> But we should force a build when it's never been built at all. Which is what we'd expect to see on a CI build, no?

A CI build should be explicit in what it is doing: `build.sh configure build test`. So we don't need to worry about that particular scenario. We should avoid implicit behaviours in CI as far as we can (this includes dependency install, builds, getting correct tool versions on PATH, etc).

But I am planning on implementing some of the checks in the build.sh refactor work I will be doing soon. There's a bit of reinvent-the-wheel in all this because that's what build orchestrators are supposed to be good at, but I have not found any build orchestrators that cover all the different build environments we work in, and a simple file existence function in build-utils.sh will probably do enough of what we need.

Long term, we may be able to use another build orchestrator, but I think that consistent build.sh bash-based scripts for now will get us a very long way forward.

Reply:
> A CI build should be explicit in what it is doing: build.sh configure build test. 

Okay then - how would you propose that the CI build for Web _explicitly_ specifies that all test dependencies are built in Web's `build.sh`, as that's the script that will trigger all these other test scripts?

Reply:
Let me get a little explicit here.

At present, all Web-related tests are triggered via a _single_ `test.sh` call:  `web/unit_tests/test.sh`.  That fans out in a dependency-based tree:

- `web/unit_tests/test.sh`
    - `common/core/web/input_processor/test.sh`
        - `common/predictive-text/unit_tests/test.sh`
          - Also requires `lm-worker`.
        - `common/core/web/keyboard_processor/test.sh`
        - Also requires `developer/js/build.sh`, `lm-worker`
- Also requires `common/core/web/tools/recorder/src/build.sh`, `lm-worker`

I get the feeling that to achieve what you say you want, we'll need to break the link between the individual `test.sh` scripts and have a 'master' test script that defines & runs the full tree?  We'd definitely want to keep said 'master' script in repo, though, as we're slowly adding more web packages, and we can't go leaving out components by accident.

Reply:
Ideally, in the end, we will have pretty much the same outcome as you currently have with `web/unit_tests/test.sh`, but we will instead call `web/build.sh [configure build] test`. (`configure` and `build` may or may not be necessary, depending on where in the CI chain it is called.) The way in which we define the hierarchy may change, but the general principle of each platform having a single script call to run all tests relevant to that platform will remain.

As to how we get there, I don't have a complete answer yet. This documentation of the very complex web of web interdependencies helps.

Some random thoughts from a tired brain:
* All scripts will be in repo. CI will only call the scripts.
* A unit test in /common having a dependency on /developer is a bit problematic (potential for circular dependencies etc) -- this is part of a larger issue where we have cross-product (particularly developer) dependencies.
* Ideally, I want each component to have its own unit tests, which can be run without reference to any other component (otherwise it is not a unit test -- it's an end-to-end test).
* Multi-module or global or end-to-end tests: I am still trying to figure out where we should put these, particularly ones that interact across multiple products.
* I think we do give each web module a `build.sh` script. This script will do `build`, preferably `test`, perhaps `configure`. It will also eventually define the module's dependencies (which we may be able to extract from tsconfig.json for web projects).
* While lerna needs to be replaced, the replacement I've chosen (npm workspaces + tsc projects) sadly duplicates dependency information in `package.json` and `tsconfig.json` 🤮. So I am seriously keen to avoid a third lot of duplication! (Weird idea from fevered imagination: what if `package.json` and/or `tsconfig.json` were generated in the `configure` step?!?)